File: aoc/nine.py
from itertools import product


class Solver:
    def __init__(self, input_file):
        self.input = self.parse(input_file)

    def parse(self, input_file):
        return [int(x) for x in open(input_file).readlines()]

    def part_one(self):
        prior = self.input[:25]
        for n in self.input[25:]:
            found = False
            for x, y in product(prior, repeat=2):
                if x + y == n:
                    found = True
                    break
            if not found:
                self.target = n
                return n
            prior = prior[1:] + [n]

    def part_two(self):
        lower, x = 0, 0
        for upper in range(len(self.input)):
            x += self.input[upper]
            while lower <= upper and x > self.target:
                x -= self.input[lower]
                lower += 1
            if x == self.target:
                contig = self.input[lower:upper]
                return min(contig) + max(contig)

    def solve(self):
        print(f"Part One: {self.part_one()}")
        s = self.part_two()
        if s is not None:
            print(f"Part Two: {s}")

File: src/nine.rs
use itertools::Itertools;

pub fn solve(input: String) {
    let data = parse(input);
    let answer_one = part_one(&data);
    println!("Part one: {}", answer_one);
    let answer_two = part_two(data, answer_one);
    println!("Part two: {}", answer_two);
}

fn parse(input: String) -> Vec<usize> {
    input.lines()
        .map(|l| l.parse().unwrap())
        .collect()
}

fn part_one(data: &Vec<usize>) -> usize {
    data.windows(26).find(
        |w| w[0..25].iter()
            .permutations(2)
            .all(|x| x[0] + x[1] != w[25])
        )
        .map(|x| x[25])
        .unwrap()
}

fn part_two(data: Vec<usize>, target: usize) -> usize {
    let aos: Vec<usize> = data.iter().scan(0, |acc, &x| {
        *acc = *acc + x;
        Some(*acc)
    }).collect();
    let mut lo = 0;
    let mut hi = 1;
    while aos[hi] - aos[lo] != target {
        if aos[hi] - aos[lo] < target {
            hi += 1;
            if hi >= aos.len() {
                panic!("upper bound exceeds array size");
            }
        } else {
            lo += 1;
        }
    }
    data[lo..hi].iter().min().unwrap() + data[lo..hi].iter().max().unwrap()
}
